src/app/model/services/login.service.ts

```diff
@@ -1,31 +1,97 @@
-import {HttpResponse, HttpHeaders} from '@angular/common/http';
+import {HttpResponse} from '@angular/common/http';
 import {Injectable} from '@angular/core';
+import * as jwt from 'jsonwebtoken';
+import * as moment from 'moment';
 import {Observable} from 'rxjs';
 import {map} from 'rxjs/operators';
 
 import {User} from '../classes/users/user';
 
 import {HttpClientService} from './http-client.service';
 
+export interface StalkerJWT {
+  organizations: {organizationId: number; role: string}[];
+  jti: string;
+  sub: string;
+  iat: number;
+  exp: number;
+}
+
 @Injectable({
   providedIn: 'root',
 })
 export class LoginService {
-  constructor(private readonly httpClientService: HttpClientService) {}
+  private readonly adminMapping = new Map();
+  constructor(private readonly httpClientService: HttpClientService) {
+    this.adminMapping.set('Admin', 4);
+    this.adminMapping.set('Owner', 3);
+    this.adminMapping.set('Manager', 2);
+    this.adminMapping.set('Viewer', 1);
+  }
 
   login(user: User): Observable<User> {
-    return this.httpClientService
-      .post<User>('/user/login', user)
-      .pipe(map((response: HttpResponse<User>) => response.body as User));
+    return this.httpClientService.post<User>('/user/login', user).pipe(
+      map((response: HttpResponse<User>) => {
+        console.log(response);
+        // let jwtTokenHeader = response.headers.get('Authorization') as string;
+        const jwtTokenHeader =
+          'Bearer eyJhbGciOiJIUzI1NiJ9.eyJvcmdhbml6YXRpb25zIjpbeyJvcmdhbml6YXRpb25JZCI6' +
+          'MSwicm9sZSI6IlZpZXdlciJ9LHsib3JnYW5pemF0aW9uSWQiOjIsInJvbGUiOiJBZG1pbiJ9XSwi' +
+          'anRpIjoiMiIsInN1YiI6Imdpb3JnaW90ZXN0MDJAaG90bWFpbC5pdCIsImlhdCI6MTU4ODMyODkz' +
+          'MSwiZXhwIjoxNTkxMzI4OTMxfQ.E7IRmte9p6-Yrl2B6iQBvQ9qxzwoCkO1lXgmjvbhlKk';
+        const jwtToken = jwtTokenHeader.substring(7);
+        const payload: StalkerJWT = jwt.verify(
+          jwtToken,
+          'OurSecureSecretLongKeyForStalkerJWTtokens',
+        ) as StalkerJWT;
+
+        console.log(payload);
+        localStorage.setItem('token', jwtToken);
+        localStorage.setItem('user_id', payload.jti);
+        localStorage.setItem('user_email', payload.sub);
+        localStorage.setItem('organizations', JSON.stringify(payload.organizations));
+        localStorage.setItem('expiration_time', payload.exp.toString());
+        localStorage.setItem('creation_time', payload.iat.toString());
```

Comment:
Ho visto che localStorage però lo posso modificare client side, avevi mai fatto caso?

Reply:
si lo so, però sembrava l'alternativa migliore perchè le altre due possibilità sono i sessionStorage che però si perdono ogni volta che chiudo il browser o i cookie che possono comunque essere modificati e in più vengono sempre trasmessi e possono essere sniffati, quindi mi sembrava la soluzione migliore perchè anche se cambia i suoi valori il token rimane invariato e il server lo sgama, solo che prima fa la richiesta pensando di avere accesso

Reply:
http://diveintohtml5.info/storage.html 
questo è uno di quelli che ho letto

Reply:
> Modern web apps include 3rd party JavaScript libraries for A/B testing, funnel/market analysis, and ads. We use package managers like Bower to import other peoples’ code into our apps.
> 
> What if only one of the scripts you use is compromised? Malicious JavaScript can be embedded on the page, and Web Storage is compromised. These types of XSS attacks can get everyone’s Web Storage that visits your site, without their knowledge. This is probably why a bunch of organizations advise not to store anything of value or trust any information in web storage. This includes session identifiers and tokens.

fonte: [link](https://stormpath.com/blog/where-to-store-your-jwts-cookies-vs-html5-web-storage)

Quindi il JWT di ogni utente messo in local storage è possibilmente accessibile da ogni libreria di terze parti che usiamo (pensiamo solo a quelle dei test) che viene compromessa, quindi l'unica vera soluzione è usare https, quindi il token che viene inviato sarà comunque criptato e reso inutilizzabile anche se venisse recuperato in un attacco

Reply:
Ma pensandoci, l'uso che verrà fatto di questo sito è un dettaglio che non dipende da noi (e quindi neanche il suo rilascio su un server con SSL), direi che usare localStorage va bene

Reply:
> quindi l'unica vera soluzione è usare https, quindi il token che viene inviato sarà comunque criptato e reso inutilizzabile anche se venisse recuperato in un attacco

occhio che questa roba è vera solo per i man in the middle, https nel momento in cui arriva sul browser viene decriptato quindi localStorage comunque contiene una copia in "chiaro" del token (che comunque è firmato da una secret key nel server)

in particolare XSS accede attraverso script iniettati nel client quindi https non è sufficiente a proteggere contro quello

Reply:
morale: localStorage va bene se usiamo CSP:

 https://angular.io/guide/security#content-security-policy

e

https://www.html5rocks.com/en/tutorials/security/content-security-policy/